File: android/src/main/kotlin/com/syncrhrohealth/bandpass_ble_connection/plugin/devicecore/method/DeviceCoreMethodPlugin.kt
package com.syncrhrohealth.bandpass_ble_connection.plugin.devicecore.method

import android.content.Context
import com.syncrhrohealth.bandpass_ble_connection.plugin.ChannelNames
import com.syncrhrohealth.bandpass_ble_connection.plugin.devicecore.enum.DeviceCoreMethod
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class DeviceCoreMethodPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var mContext: Context
    private val methodChannel = ChannelNames.DEVICE_CORE_METHOD_CHANNEL

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        try {
            mContext = binding.applicationContext
            val channel = MethodChannel(binding.binaryMessenger, methodChannel)
            channel.setMethodCallHandler(this)
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments<ArrayList<*>>()

        when (call.method) {
            DeviceCoreMethod.IS_BLE_ENABLED.value -> {
                DeviceCoreMethodHelper.getInstance(mContext).isBleEnabled(args, result)
            }

            DeviceCoreMethod.START_SCAN.value -> {
                DeviceCoreMethodHelper.getInstance(mContext).startScan(args, result)
            }

            DeviceCoreMethod.STOP_SCAN.value -> {
                DeviceCoreMethodHelper.getInstance(mContext).stopScan(args, result)
            }

            DeviceCoreMethod.CONNECT.value -> {
                DeviceCoreMethodHelper.getInstance(mContext).connect(args, result)
            }

            DeviceCoreMethod.DISCONNECT.value -> {
                DeviceCoreMethodHelper.getInstance(mContext).disconnect(args, result)
            }

            DeviceCoreMethod.RECONNECT.value -> {
                DeviceCoreMethodHelper.getInstance(mContext).reconnect(args, result)
            }

            DeviceCoreMethod.RECONNECT_DEVICES.value -> {
                // TODO: Implement
            }

            DeviceCoreMethod.GET_DEVICE_INFO.value -> {
                DeviceCoreMethodHelper.getInstance(mContext).getDeviceInfo(args, result)
            }

            else -> result.notImplemented()
        }
    }


    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        //NO IMPLEMENTATION
    }
}

File: ios/Classes/Plugin/DeviceCore/Method/DeviceCoreMethodPlugin.swift
//
//  DeviceCoreMethodPlugin.swift
//  Runner
//
//  Created by MAC on 10/3/25.
//

import Foundation
import Flutter

class DeviceCoreMethodPlugin: NSObject {
    
    private let eventHandler = DeviceCoreEventHandler()
    
    static func register(with messenger: FlutterBinaryMessenger){
        let instance = DeviceCoreMethodPlugin()
        
        let channel = FlutterMethodChannel(name: ChannelNames.DEVICE_CORE_METHOD_CHANNEL, binaryMessenger: messenger)
        channel.setMethodCallHandler(instance.handle)
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let params = call.arguments as? Array<Any>
        NSLog("[DeviceCoreMethodPlugin] handle: \(call.method)")
//        FileLogger.log("[CorePlugin] - handle: \(call.method)")
        switch call.method {
        case DeviceCoreMethod.hasBlePermission.rawValue:
            DeviceCoreMethodHelper.shared.hasBlePermission(params: params, result: result)
            break
        case DeviceCoreMethod.isBleEnabled.rawValue:
            DeviceCoreMethodHelper.shared.isBleEnabled(params: params, result: result)
            break
        case DeviceCoreMethod.startScan.rawValue:
            DeviceCoreMethodHelper.shared.startScan(params: params, result: result)
            break
        case DeviceCoreMethod.stopScan.rawValue:
            DeviceCoreMethodHelper.shared.stopScan(params: params, result: result)
            break
        case DeviceCoreMethod.connect.rawValue:
            DeviceCoreMethodHelper.shared.connect(params: params, result: result)
            break
        case DeviceCoreMethod.disconnect.rawValue:
            DeviceCoreMethodHelper.shared.disconnect(params: params, result: result)
            break
        case DeviceCoreMethod.reconnect.rawValue:
            DeviceCoreMethodHelper.shared.reconnect(params: params, result: result)
            break
        case DeviceCoreMethod.reconnectDevices.rawValue:
            // TODO: Implement reconnectDevices
            result(true)
            break
            
        case DeviceCoreMethod.getDeviceInfo.rawValue:
            DeviceCoreMethodHelper.shared.getDeviceInfo(params: params, result: result)
            break
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
}
